c/src/ml-api-service-db.cc

```diff
@@ -0,0 +1,268 @@
+ * @author Sangjung Woo <sangjung.woo@samsung.com>
+ * @bug No known bugs except for NYI items
+ */
+
+#include <iostream>
+#include <cstring>
+
+#include "ml-api-inference-internal.h"
+#include "ml-service-db.hh"
+#include "ml-api-service.h"
+
+#define ML_DATABASE_PATH      SYS_DB_DIR"/.ml-service-leveldb"
+
+/**
+ * @brief Class for implementation of IMLServiceDB
+ */
+class MLServiceLevelDB : public IMLServiceDB
+{
+public:
+  MLServiceLevelDB (std::string path);
+  virtual ~ MLServiceLevelDB ();
+  virtual void connectDB () override;
+  virtual void setPipelineDescription (std::string name,
+      std::string pipeline_description) override;
+  virtual void getPipelineDescription (std::string name,
+      std::string & pipeline_description) override;
+  virtual void delPipelineDescription (std::string name);
+
+private:
+  std::string path;
+  leveldb_t *db_obj;
+  leveldb_readoptions_t *db_roptions;
+  leveldb_writeoptions_t *db_woptions;
+};
+
+static IMLServiceDB *instance = nullptr;
+
+/**
+ * @brief Get the instance object of MLServiceLevelDB
+ * @return IMLServiceDB* instance of MLServiceLevelDB
+ */
+static IMLServiceDB *
+get_instance ()
+{
+  if (!instance) {
+    instance = new MLServiceLevelDB (ML_DATABASE_PATH);
+    instance->connectDB ();
+  }
+
+  return instance;
+}
+
+/**
+ * @brief Construct a new MLServiceLevelDB object
+ * @param path database path
+ */
+MLServiceLevelDB::MLServiceLevelDB (std::string path)
+:  path (path)
+{
+  db_obj = nullptr;
+}
+
+/**
+ * @brief Destroy the MLServiceLevelDB object
+ */
+MLServiceLevelDB::~MLServiceLevelDB ()
+{
+}
+
+/**
+ * @brief Connect the level DB and initialize the private variables.
+ */
+void
+MLServiceLevelDB::connectDB ()
+{
+  char *err = nullptr;
+  leveldb_options_t *db_options;
+
+  db_options = leveldb_options_create ();
+  leveldb_options_set_create_if_missing (db_options, 1);
+
+  db_obj = leveldb_open (db_options, path.c_str (), &err);
+  if (err != nullptr) {
+    _ml_loge ("Error! Failed to open Database : %s", err);
+    leveldb_free (err);
+    throw std::runtime_error ("Failed to connectDB()!");
+  }
+
+  db_roptions = leveldb_readoptions_create ();
+  db_woptions = leveldb_writeoptions_create ();
+  leveldb_writeoptions_set_sync (db_woptions, 1);
+
+  return;
+}
+
+/**
+ * @brief Set the pipeline description with the given name.
+ * @note If the name already exists, the pipeline description is overwritten.
+ * @param[in] name Unique name to retrieve the associated pipeline description.
+ * @param[in] pipeline_description The pipeline description to be stored.
+ */
+void
+MLServiceLevelDB::setPipelineDescription (std::string name,
+    std::string pipeline_description)
+{
+  char *err = NULL;
+
+  leveldb_put (db_obj, db_woptions, name.c_str (), name.size (),
+      pipeline_description.c_str (), pipeline_description.size (), &err);
+  if (err != nullptr) {
+    _ml_loge ("Failed to put %s : %s", name.c_str (), err);
```

Comment:
```
"Failed to call leveldb_put () for the name, '%s' of the pipeline description (size: %d bytes / decription: '%.40s%s'). leveldb_put () has returned an error: %s", ... , (name.size() > 40) ? "...", ""
```

Try to give as much info as possible that may be used for debugging.

Reply:
I will.